### PR TITLE
ISPN-6504 Try to use the index even if the query matches all entries

### PR DESCRIPTION
**Pull request by anistor**
https://issues.jboss.org/browse/ISPN-6504
https://issues.jboss.org/browse/ISPN-6506

### CODE REVIEW

File: core/src/main/java/org/infinispan/notifications/cachelistener/CacheNotifierImpl.java [hunk 1]
```diff
@@ -82,9 +81,8 @@
    private static final Log log = LogFactory.getLog(CacheNotifierImpl.class);
    private static final boolean trace = log.isTraceEnabled();
 
-   private static final Map<Class<? extends Annotation>, Class<?>> allowedListeners = new HashMap<Class<? extends Annotation>, Class<?>>(16);
-   private static final Map<Class<? extends Annotation>, Class<?>> clusterAllowedListeners =
-         new HashMap<Class<? extends Annotation>, Class<?>>(4);
+   private static final Map<Class<? extends Annotation>, Class<?>> allowedListeners = new HashMap<>(16);
```
Comment (gustavonalle) [inline core/src/main/java/org/infinispan/notifications/cachelistener/CacheNotifierImpl.java]: "Apply Java 7 syntax" you probably meant in the commit?

File: object-filter/src/main/java/org/infinispan/objectfilter/impl/syntax/ConstantValueExpr.java [hunk 2]
```diff
@@ -35,7 +35,7 @@ public String toString() {
 
       @Override
       public int compareTo(Object o) {
-         return 0;  //todo [anistor] hack!
+         return name.compareTo(((ParamPlaceholder) o).getName());
```
Comment (gustavonalle) [inline object-filter/src/main/java/org/infinispan/objectfilter/impl/syntax/ConstantValueExpr.java]: null check here?

Comment (anistor) [inline object-filter/src/main/java/org/infinispan/objectfilter/impl/syntax/ConstantValueExpr.java]: The contract for compareTo allows failing with NullPointerException is arg is null

File: query/src/main/java/org/infinispan/query/dsl/embedded/impl/QueryEngine.java [hunk 3]
```diff
@@ -737,20 +737,27 @@ protected CacheQuery buildLuceneQuery(FilterParsingResult<?> filterParsingResult
...
-      HibernateSearchPropertyHelper propertyHelper = new HibernateSearchPropertyHelper(getSearchFactory(), entityNamesResolver, null);
-      return JPALuceneTransformer.transform(parsingResult, getSearchFactory(), entityNamesResolver, propertyHelper::getDefaultFieldBridge, namedParameters);
+   private <TypeMetadata> LuceneQueryParsingResult<TypeMetadata> transform(FilterParsingResult<TypeMetadata> parsingResult, Map<String, Object> namedParameters) {
+      LuceneQueryParsingResult<TypeMetadata> luceneParsingResult;
+      if (queryCache != null && parsingResult.getParameterNames().isEmpty()) {
+         KeyValuePair<String, Class> queryCacheKey = new KeyValuePair<>(parsingResult.getJpaQuery(), LuceneQueryParsingResult.class);
+         luceneParsingResult = queryCache.get(queryCacheKey);
+         if (luceneParsingResult == null) {
+            luceneParsingResult = createLuceneMaker().transform(parsingResult, namedParameters, getTargetedClass(parsingResult));
+            queryCache.put(queryCacheKey, luceneParsingResult);
```
Comment (gustavonalle) [inline query/src/main/java/org/infinispan/query/dsl/embedded/impl/QueryEngine.java]: Lucene queries are not serializable, we have externalizers for some of them, I suppose all the queries generated by ```createLuceneMaker``` are covered?

File: object-filter/src/main/java/org/infinispan/objectfilter/impl/hql/predicate/FilterPredicateFactory.java [hunk 4]
```diff
@@ -25,18 +24,15 @@
...
-      this.entityNamesResolver = entityNamesResolver;
+   public FilterPredicateFactory(ObjectPropertyHelper propertyHelper) {
       this.propertyHelper = propertyHelper;
    }
 
    @Override
    public RootPredicate<BooleanExpr> getRootPredicate(String entityType) {
-      if (entityNamesResolver.getClassFromName(entityType) == null) {
+      if (propertyHelper.getEntityNamesResolver().getClassFromName(entityType) == null) {
          throw new IllegalStateException("Unknown entity name " + entityType);
```
Comment (gustavonalle) [inline object-filter/src/main/java/org/infinispan/objectfilter/impl/hql/predicate/FilterPredicateFactory.java]: ```throw log.getNoIndexedEntityException(entityType)```  ?

Comment (gustavonalle) [inline object-filter/src/main/java/org/infinispan/objectfilter/impl/hql/predicate/FilterPredicateFactory.java]: or something similar  :smile:

File: core/src/main/java/org/infinispan/notifications/cachelistener/CacheNotifierImpl.java (hunk 1 above)
Comment (anistor) [inline core/src/main/java/org/infinispan/notifications/cachelistener/CacheNotifierImpl.java]: Java 8 includes Java 7 :)

File: query/src/main/java/org/infinispan/query/dsl/embedded/impl/QueryEngine.java (hunk 3 above)
Comment (anistor) [inline query/src/main/java/org/infinispan/query/dsl/embedded/impl/QueryEngine.java]: The query cache is not replicated, so this should not be a concern.

File: object-filter/src/main/java/org/infinispan/objectfilter/impl/hql/predicate/FilterPredicateFactory.java (hunk 4 above)
Comment (anistor) [inline object-filter/src/main/java/org/infinispan/objectfilter/impl/hql/predicate/FilterPredicateFactory.java]: Indeed, I'll do that.

Comment (anistor) [inline object-filter/src/main/java/org/infinispan/objectfilter/impl/hql/predicate/FilterPredicateFactory.java]: Done.

**Comment by gustavonalle** [comment]
pulling

**Comment by gustavonalle** [comment]
merged, thanks @anistor

**Pull request merged**